File: src/triangle.cpp
#include "drawable.h"

class Triangle: public Drawable{
protected:
	float vertices[] = {
	    // positions         // colors
	     0.5f, -0.5f, 0.0f,  1.0f, 0.0f, 0.0f,   // bottom right
	    -0.5f, -0.5f, 0.0f,  0.0f, 1.0f, 0.0f,   // bottom left
	     0.0f,  0.5f, 0.0f,  0.0f, 0.0f, 1.0f    // top 
	};
public:
	Drawable(const char* vert, const char* frag){
		GLuint VBO;
	
		glGenVertexArrays(1, &VAO);
		glBindVertexArray(VAO);
		glGenBuffers(1, &VBO);
	
		glBindBuffer(GL_ARRAY_BUFFER, VBO);
		glBufferData(GL_ARRAY_BUFFER, sizeof(this->vertices), this->vertices, GL_STATIC_DRAW);
	
		//Use layout 0, vec3, float, don't normalize, how large is each component, pointer to offset.
		glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 6*sizeof(GL_FLOAT), (void*)0);
		glEnableVertexAttribArray(0);
	
		glVertexAttribPointer(1, 3, GL_FLOAT, GL_FALSE, 6*sizeof(GL_FLOAT), (void*)(3*sizeof(float)));
		glEnableVertexAttribArray(1);
	}
	~Drawable(){
		
	}
	void DrawShape(){
		
	}
};

File: graphics/Triangle.java
package graphics;

import static org.lwjgl.opengl.GL33.*;
import static org.lwjgl.system.MemoryUtil.*;
import static org.lwjgl.stb.STBImage.*;

import java.nio.IntBuffer;
import java.nio.ByteBuffer;
import org.lwjgl.system.MemoryStack;

public class Triangle extends Drawable{

	private static final float vertexData[] = {
		    // positions		Colors
		     0.5f, -0.5f, 0.0f, 1.0f, 0.0f, 0.0f,
		    -0.5f, -0.5f, 0.0f, 0.0f, 1.0f, 0.0f,
		     0.0f,  0.5f, 0.0f, 0.0f, 0.0f, 1.0f
	};
	
	public Triangle() {
		try {
			String vert = getClass().getClassLoader().getResource("resources/shaders/triangle.vert").getPath();
			String frag = getClass().getClassLoader().getResource("resources/shaders/triangle.frag").getPath();
			
			shader = new Shader(vert, frag);
		}catch(NullPointerException e) {
			System.err.println("Cannot find the resource in the path. Please make sure there isn't a typo in the shader you are requesting.");
			System.exit(-1);
		}
				
		glBindVertexArray(VAO);
	
		glBindBuffer(GL_ARRAY_BUFFER, VBO);
		glBufferData(GL_ARRAY_BUFFER, vertexData, GL_STATIC_DRAW);
		
		int aPos = glGetAttribLocation(shader.GetProgramID(), "aPos");
		int aColor = glGetAttribLocation(shader.GetProgramID(), "aColor");
		
		glVertexAttribPointer(aPos, 3, GL_FLOAT, false, 6*Float.BYTES, NULL);
		glEnableVertexAttribArray(aPos);
		
		glVertexAttribPointer(aColor, 3, GL_FLOAT, false, 6*Float.BYTES, 3*Float.BYTES);
		glEnableVertexAttribArray(aColor);
		
		glBindBuffer(GL_ARRAY_BUFFER, 0);
		glBindVertexArray(0);
	}
	
	public void SetShader(Shader shader) {
		this.shader = shader;
	}
	
	public int GetVAO() {
		return VAO;
	}
	
	@Override
	public void Render() {
		shader.Use();
		glBindVertexArray(this.VAO);
		glDrawArrays(GL_TRIANGLES, 0, 3);	
	}
}
